clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java

```diff
@@ -248,7 +248,10 @@ public synchronized ConsumerRecords<K, V> poll(final Duration timeout) {
                         subscriptions.position(entry.getKey(), newPosition);
                     }
                 }
-                toClear.add(entry.getKey());
+                // Since reassignment logic for tiered topics relies on seekToBeginning,
```

Comment:
I don't think we can change this logic, because it would re-deliver the same records over and over, which is a distinct from the real KafkaConsumer that (without seeking) tries to only deliver records once.

I do think that maybe seekToBeginning should clear the records from the reset partitions, to mirror how the real kafka consumer uses FetchState.AWAIT_RESET#hasValidPosition to prevent fetching data from a partition which has just been reset, even if there is data buffered.

Then the test can explicitly re-add the records to re-consume them.